Feat: Keyring entity (#243)
Kong/developer.konghq.com: app/_gateway_entities/keyring.md

## Encryption details

### Encrypted fields

The Keyring encrypts the following fields:

- The `key` fields in a `certificate` object, which is the private key for a TLS certificate.
- Certain configuration parameters in plugins and plugin-related authentication objects. These parameters have an `encrypted` label in the plugin's configuration reference.

### Encryption and decryption

{{site.base_gateway}} uses 256-bit AES encryption in GCM mode. Cryptographic nonce values for each encryption routine execution are derived from the kernel CSPRNG (`/dev/urandom`). The AES routines used by Kong are provided by the OpenSSL library bundled with the {{site.base_gateway}} package.

## Key generation and lifecycle

{{site.base_gateway}}’s keyring handling mechanisms allow for more than one key to be present on any given Kong node at a time. Each key may be used to read encrypted fields from the database, but only one key at any given time is used to write encrypted fields back to the data store. This process allows for a key rotation mechanism wherein new keyring material is introduced, and older keys may be present for a time to allow rotating previously-encrypted fields.

> Through the kernel CSPRNG, Kong derives keyring material generated through the `/keyring/generate` Admin API endpoint. Kong stores keyring material in a shared memory zone that all Kong worker processes access. To prevent key material from being written to disk as part of memory paging operations, we recommend that swap be disabled on systems running Kong.

Author: The swap refers to a parameter on the machine running Kong, not in Kong, so there's not endpoint to provide

> When operating in cluster mode, keyring material propagates automatically among all nodes in the Kong cluster. Because Kong nodes don't have a notion of direct peer-to-peer communication, the underlying data store serves as a communication channel to transmit messages. When a Kong node starts, it generates an ephemeral RSA key pair. The node’s public keys propagate to all other active nodes in the cluster. When an active node sees a message request for keyring material, it wraps the in-memory keyring material in the presented public key, and transmits the payload back over the central messaging channel provided by the underlying data store. This process allows each node in the cluster to broadcast keyring material to new nodes, without sending key material in plain text over the wire. This model requires that at least one node be running at all times within the cluster; a failure of all nodes requires manually re-importing the keyring to one node during an outage recovery.

Reviewer: I'd like a diagram or a list that explicitly states what happens when keyring is enabled and data comes in that needs to be encrypted. It's difficult for me to process how keyring works with the long paragraphs of text.

What I'd like to see covered:
- Read/write encryption flow between the database
- how encrypted data is broadcasted to nodes
- How new/old keyring material is handled

## Disaster recovery

> To automatically back up your Keyring material, make sure to configure the `keyring_recovery_public_key` parameter to point to your `cert.pem` file. See [Enable Keyring](#enable-keyring) for more details.

Reviewer: ```suggestion
To automatically back up your Keyring-encrypted data, make sure to configure the `keyring_recovery_public_key` parameter to point to your `cert.pem` file. See [Enable Keyring](#enable-keyring) for more details.
```
I think this is more accurate, but I'm not sure. Since Keyring is a mechanism to encrypt data, it didn't feel like Keyring could have material.

Author: the keyring material refers to the keys in the keyring

> The Keyring material is then encrypted with the public RSA key defined with the `keyring_recovery_public_key` value in the database. The corresponding private key can be used to decrypt the keyring material in the database:

Reviewer: ```suggestion
The Keyring data is then encrypted with the public RSA key defined with the `keyring_recovery_public_key` value in the database. The corresponding private key can be used to recover keys by decrypting the keyring material in the database:
```
I don't understand what "keyring material" is. I think it might be unencrypted fields that keyring will eventually encrypt. Please define "keyring material" on the first instance on this page.

```sh
curl -X POST localhost:8001/keyring/recover -F recovery_private_key=@/path/to/generated/key.pem
```

The response contains a list of keys that were successfully recovered and a list of keys that could not be recovered. The Kong error log will contain the detailed reason why the keys could not be recovered.

## Enable Keyring

### 1. Generate an RSA key pair

> Use the `openssl` CLI to generate an RSA key pair that can be used to export and recover keyring material.

Reviewer: ```suggestion
Use the `openssl` CLI to generate an RSA key pair that can be used to export and recover keyring material:
```

```sh
openssl genrsa -out key.pem 2048
openssl rsa -in key.pem -pubout -out cert.pem
```

### 2. Configure Kong

> Enabling data encryption in Kong requires modifying the Kong configuration. You can either update `kong.conf` or use environment variables:

Reviewer: ```suggestion
To enable data encryption, you must modify the Kong configuration. You can either update `kong.conf` or use environment variables:
```

{% navtabs %}
{% navtab "kong.conf" %}
Set the following values in `kong.conf`:
```sh
keyring_enabled = on
keyring_strategy = cluster
keyring_recovery_public_key = /path/to/generated/cert.pem
```
{% endnavtab %}
{% navtab "Environment variables" %}
Create the following environment variables:
```sh
export KONG_KEYRING_ENABLED=on
export KONG_KEYRING_STRATEGY=cluster
export KONG_KEYRING_RECOVERY_PUBLIC_KEY=/path/to/generated/cert.pem
```
{% endnavtab %}
{% endnavtabs %}

### 3. Start Kong and validate
Once the configuration is updated, you can start your {{site.base_gateway}} instance and use the following request to make sure the Keyring is enabled:

```sh

> curl -s localhost:8001/keyring

Reviewer: I tested with the environment variable flow and it worked. I tried doing the kong.conf flow on a new Gateway quickstart install, and for some reason, I kept getting  `{"ids":{}}` before I was able to add the values to kong.conf.

Author: Ah you're right, I'm getting the same. After a bit of testing, it looks like the only way to really know if it's working is to look directly in the database to see if the data is encrypted. I'm going to update this part.
